File: src/test/java/com/gemini/krakenbot/KrakenE2ETest.java
package com.gemini.krakenbot;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.gemini.krakenbot.config.AppConfig;
import com.gemini.krakenbot.config.Allocation;
import com.gemini.krakenbot.config.KrakenCredentials;
import com.gemini.krakenbot.config.Settings;
import com.gemini.krakenbot.model.PortfolioStats;
import com.gemini.krakenbot.repository.PortfolioStatsRepository;
import com.gemini.krakenbot.service.ConfigService;
import com.gemini.krakenbot.service.TradeHistoryService;
import com.gemini.krakenbot.service.impl.KrakenServiceImpl;
import com.gemini.krakenbot.service.impl.PortfolioManagerImpl;
import org.junit.jupiter.api.Test;
import org.springframework.http.HttpMethod;
import org.springframework.http.MediaType;
import org.springframework.test.web.client.MockRestServiceServer;
import org.springframework.web.client.RestClient;

import java.math.BigDecimal;
import java.util.Base64;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;
import static org.springframework.test.web.client.match.MockRestRequestMatchers.method;
import static org.springframework.test.web.client.match.MockRestRequestMatchers.requestTo;
import static org.springframework.test.web.client.match.MockRestRequestMatchers.requestToUriTemplate;
import static org.springframework.test.web.client.response.MockRestResponseCreators.withSuccess;

public class KrakenE2ETest {

    private ObjectMapper objectMapper = new ObjectMapper().registerModule(new JavaTimeModule());

    @Test
    void shouldExecuteFullRebalanceCycleEndToEndWithoutOrder() throws Exception {
        String validSecret = Base64.getEncoder().encodeToString("secret".getBytes());
        AppConfig appConfig = new AppConfig(
                new KrakenCredentials("apiKey", validSecret),
                new Settings(60L, 2.0, 1.0, false, 50.0, 1.0),
                List.of(new Allocation("BTC", 50.0), new Allocation("USD", 50.0))
        );

        ConfigService mockConfigService = mock(ConfigService.class);
        when(mockConfigService.getConfig()).thenReturn(appConfig);

        RestClient.Builder builder = RestClient.builder();
        MockRestServiceServer mockServer = MockRestServiceServer.bindTo(builder).build();

        mockServer.expect(requestTo("https://api.kraken.com/0/private/Balance"))
                .andExpect(method(HttpMethod.POST))
                .andRespond(withSuccess("{\"error\":[],\"result\":{\"XXBT\":0.5,\"ZUSD\":25000.0}}", MediaType.APPLICATION_JSON));

        mockServer.expect(requestTo("https://api.kraken.com/0/public/Ticker?pair=XBTUSD"))
                .andExpect(method(HttpMethod.GET))
                .andRespond(withSuccess("{\"error\":[],\"result\":{\"XXBTZUSD\":{\"c\":[\"50000.0\"]}}}", MediaType.APPLICATION_JSON));

        PortfolioStatsRepository mockStatsRepo = mock(PortfolioStatsRepository.class);
        when(mockStatsRepo.load()).thenReturn(new PortfolioStats(BigDecimal.ZERO));

        KrakenServiceImpl krakenService = new KrakenServiceImpl(mockConfigService, objectMapper, builder);
        PortfolioManagerImpl portfolioManager = new PortfolioManagerImpl(krakenService, mockConfigService, mock(TradeHistoryService.class), mockStatsRepo);

        invokePerformRebalanceCycle(portfolioManager);

        // Verify that no order was placed (MockRestServiceServer will throw an AssertionError if AddOrder was called without being expected)
        mockServer.verify();
    }

    @Test
    void shouldExecuteFullRebalanceCycleEndToEndAndTriggerTrade() throws Exception {
        String validSecret = Base64.getEncoder().encodeToString("secret".getBytes());
        AppConfig appConfig = new AppConfig(
                new KrakenCredentials("apiKey", validSecret),
                new Settings(60L, 2.0, 1.0, false, 50.0, 1.0),
                List.of(new Allocation("BTC", 50.0), new Allocation("USD", 50.0))
        );

        ConfigService mockConfigService = mock(ConfigService.class);
        when(mockConfigService.getConfig()).thenReturn(appConfig);

        RestClient.Builder builder = RestClient.builder();
        MockRestServiceServer mockServer = MockRestServiceServer.bindTo(builder).build();

        mockServer.expect(requestTo("https://api.kraken.com/0/private/Balance"))
                .andExpect(method(HttpMethod.POST))
                .andRespond(withSuccess("{\"error\":[],\"result\":{\"XXBT\":0.4,\"ZUSD\":30000.0}}", MediaType.APPLICATION_JSON));

        mockServer.expect(requestTo("https://api.kraken.com/0/public/Ticker?pair=XBTUSD"))
                .andExpect(method(HttpMethod.GET))
                .andRespond(withSuccess("{\"error\":[],\"result\":{\"XXBTZUSD\":{\"c\":[\"50000.0\"]}}}", MediaType.APPLICATION_JSON));

        // We use string matchers or regex to assert the request body contains volume=0.1
        mockServer.expect(requestTo("https://api.kraken.com/0/private/AddOrder"))
                .andExpect(method(HttpMethod.POST))
                .andExpect(request -> {
                    String body = request.getBody().toString();
                    assertTrue(body.contains("pair=BTCUSD"), "Missing pair=BTCUSD");
                    assertTrue(body.contains("type=buy"), "Missing type=buy");
                    assertTrue(body.contains("ordertype=market"), "Missing ordertype=market");
                    assertTrue(body.contains("volume=0.1"), "Missing volume=0.1");
                })
                .andRespond(withSuccess("{\"error\":[],\"result\":{\"descr\":{\"order\":\"buy\"},\"txid\":[\"TX-1\"]}}", MediaType.APPLICATION_JSON));

        PortfolioStatsRepository mockStatsRepo = mock(PortfolioStatsRepository.class);
        when(mockStatsRepo.load()).thenReturn(new PortfolioStats(BigDecimal.ZERO));

        KrakenServiceImpl krakenService = new KrakenServiceImpl(mockConfigService, objectMapper, builder);
        PortfolioManagerImpl portfolioManager = new PortfolioManagerImpl(krakenService, mockConfigService, mock(TradeHistoryService.class), mockStatsRepo);

        invokePerformRebalanceCycle(portfolioManager);

        mockServer.verify();
    }

    private void invokePerformRebalanceCycle(PortfolioManagerImpl pm) throws Exception {
        java.lang.reflect.Method method = PortfolioManagerImpl.class.getDeclaredMethod("performRebalanceCycle");
        method.setAccessible(true);
        try {
            method.invoke(pm);
        } catch (java.lang.reflect.InvocationTargetException e) {
            throw (Exception) e.getCause();
        }
    }
}

File: src/test/kotlin/com/gemini/krakenbot/KrakenE2ETest.kt
package com.gemini.krakenbot

import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import com.gemini.krakenbot.config.Allocation
import com.gemini.krakenbot.config.AppConfig
import com.gemini.krakenbot.config.DatabaseConfig
import com.gemini.krakenbot.config.KrakenCredentials
import com.gemini.krakenbot.config.Settings
import com.gemini.krakenbot.model.Asset
import com.gemini.krakenbot.repository.impl.SqlitePortfolioStatsRepositoryImpl
import com.gemini.krakenbot.repository.impl.SqliteTradeRepositoryImpl
import com.gemini.krakenbot.service.ConfigService
import com.gemini.krakenbot.service.impl.ConfigServiceImpl
import com.gemini.krakenbot.service.impl.KrakenServiceImpl
import com.gemini.krakenbot.service.impl.OrderExecutorImpl
import com.gemini.krakenbot.service.impl.PortfolioAnalyzerImpl
import com.gemini.krakenbot.service.impl.PortfolioManagerImpl
import com.gemini.krakenbot.service.impl.history.TradeHistoryServiceImpl
import io.kotest.core.spec.IsolationMode
import io.kotest.core.spec.style.StringSpec
import io.kotest.matchers.booleans.shouldBeTrue
import io.kotest.matchers.nulls.shouldBeNull
import io.kotest.matchers.nulls.shouldNotBeNull
import io.ktor.client.HttpClient
import io.ktor.client.engine.mock.MockEngine
import io.ktor.client.engine.mock.respond
import io.ktor.http.HttpHeaders
import io.ktor.http.HttpStatusCode
import io.ktor.http.content.TextContent
import io.ktor.http.headersOf
import io.mockk.every
import io.mockk.mockk
import kotlinx.coroutines.test.runTest
import java.util.Base64

class KrakenE2ETest : StringSpec() {
    override fun isolationMode() = IsolationMode.InstancePerTest

    init {
        "should execute a full rebalance cycle end-to-end" {
            runTest {
                val validSecret =
                    Base64
                        .getEncoder()
                        .encodeToString("secret".toByteArray())
                val appConfig =
                    AppConfig(
                        kraken =
                        KrakenCredentials(
                            apiKey = "apiKey",
                            privateKey = validSecret,
                        ),
                        settings =
                        TestFixtures.settings(dryRun = false, loopDelaySeconds = 60L, fiatMaxDrawdown = 50.0),
                        allocations =
                        listOf(
                            Allocation(Asset.BTC, 50.0),
                            Allocation(Asset.USD, 50.0),
                        ),
                    )

                val mockConfigService = mockk<ConfigService>(relaxed = true)
                every { mockConfigService.getConfig() } returns appConfig

                var capturedOrderPayload: String? = null

                val mockEngine =
                    MockEngine { request ->
                        when (request.url.encodedPath) {
                            "/0/private/Balance" -> {
                                respond(
                                    content = "{\"error\":[],\"result\":{\"XXBT\":0.5,\"ZUSD\":25000.0}}",
                                    status = HttpStatusCode.OK,
                                    headers =
                                    headersOf(
                                        HttpHeaders.ContentType,
                                        TestFixtures.APPLICATION_JSON,
                                    ),
                                )
                            }

                            "/0/public/Ticker" -> {
                                respond(
                                    content = "{\"error\":[],\"result\":{\"XXBTZUSD\":{\"c\":[\"50000.0\"]}}}",
                                    status = HttpStatusCode.OK,
                                    headers =
                                    headersOf(
                                        HttpHeaders.ContentType,
                                        TestFixtures.APPLICATION_JSON,
                                    ),
                                )
                            }

                            "/0/private/AddOrder" -> {
                                capturedOrderPayload =
                                    (request.body as TextContent).text
                                respond(
                                    content =
                                    "{\"error\":[],\"result\":{\"descr\":{\"order\":\"buy\"},\"txid\":[\"TX-1\"]}}",
                                    status = HttpStatusCode.OK,
                                    headers =
                                    headersOf(
                                        HttpHeaders.ContentType,
                                        TestFixtures.APPLICATION_JSON,
                                    ),
                                )
                            }

                            else -> {
                                respond(
                                    "{\"error\":[\"Unknown path\"]}",
                                    HttpStatusCode.NotFound,
                                )
                            }
                        }
                    }

                val httpClient = HttpClient(mockEngine)
                val objectMapper =
                    jacksonObjectMapper().findAndRegisterModules()

                val db = DatabaseConfig.init(":memory:")
                val statsRepo = SqlitePortfolioStatsRepositoryImpl(db, objectMapper)
                val tradesRepo = SqliteTradeRepositoryImpl(db)

                val krakenService =
                    KrakenServiceImpl(
                        configService = mockConfigService,
                        objectMapper = objectMapper,
                        httpClient = httpClient,
                    )
                val portfolioAnalyzer =
                    PortfolioAnalyzerImpl(
                        krakenService = krakenService,
                        configService = mockConfigService,
                        portfolioStatsRepository = statsRepo,
                    )
                val tradeHistoryService =
                    TradeHistoryServiceImpl(
                        repository = tradesRepo,
                        portfolioStatsRepository = statsRepo,
                        krakenService = krakenService,
                        configService = mockConfigService,
                        objectMapper = objectMapper,
                        portfolioAnalyzer = portfolioAnalyzer,
                    )

                val orderExecutor =
                    OrderExecutorImpl(krakenService, tradeHistoryService)
                val portfolioManager =
                    PortfolioManagerImpl(
                        configService = mockConfigService,
                        tradeHistoryService = tradeHistoryService,
                        portfolioAnalyzer = portfolioAnalyzer,
                        orderExecutor = orderExecutor,
                    )

                portfolioManager.performRebalanceCycle()

                // 0.5 BTC @ $50,000 is $25,000 against $25,000 of cash — exactly the 50/50 target,
                // so the cycle must never reach AddOrder.
                capturedOrderPayload.shouldBeNull()
            }
        }

        "should execute a full rebalance cycle end-to-end and trigger a trade" {
            runTest {
                val validSecret =
                    Base64.getEncoder().encodeToString("secret".toByteArray())
                val appConfig =
                    AppConfig(
                        kraken =
                        KrakenCredentials(
                            apiKey = "apiKey",
                            privateKey = validSecret,
                        ),
                        settings =
                        TestFixtures.settings(dryRun = false, loopDelaySeconds = 60L, fiatMaxDrawdown = 50.0),
                        allocations =
                        listOf(
                            Allocation(Asset.BTC, 50.0),
                            Allocation(Asset.USD, 50.0),
                        ),
                    )

                val mockConfigService = mockk<ConfigService>(relaxed = true)
                every { mockConfigService.getConfig() } returns appConfig

                var capturedOrderPayload: String? = null

                val mockEngine =
                    MockEngine { request ->
                        when (request.url.encodedPath) {
                            "/0/private/Balance" -> {
                                respond(
                                    content = "{\"error\":[],\"result\":{\"XXBT\":0.4,\"ZUSD\":30000.0}}",
                                    status = HttpStatusCode.OK,
                                    headers =
                                    headersOf(
                                        HttpHeaders.ContentType,
                                        TestFixtures.APPLICATION_JSON,
                                    ),
                                )
                            }

                            "/0/public/Ticker" -> {
                                respond(
                                    content = "{\"error\":[],\"result\":{\"XXBTZUSD\":{\"c\":[\"50000.0\"]}}}",
                                    status = HttpStatusCode.OK,
                                    headers =
                                    headersOf(
                                        HttpHeaders.ContentType,
                                        TestFixtures.APPLICATION_JSON,
                                    ),
                                )
                            }

                            "/0/private/AddOrder" -> {
                                capturedOrderPayload =
                                    (request.body as TextContent).text
                                respond(
                                    content =
                                    "{\"error\":[],\"result\":{\"descr\":{\"order\":\"buy\"},\"txid\":[\"TX-1\"]}}",
                                    status = HttpStatusCode.OK,
                                    headers =
                                    headersOf(
                                        HttpHeaders.ContentType,
                                        TestFixtures.APPLICATION_JSON,
                                    ),
                                )
                            }

                            else ->
                                respond(
                                    "{\"error\":[\"Unknown path\"]}",
                                    HttpStatusCode.NotFound,
                                )
                        }
                    }

                val httpClient = HttpClient(mockEngine)
                val objectMapper =
                    jacksonObjectMapper().findAndRegisterModules()

                val db = DatabaseConfig.init(":memory:")
                val statsRepo = SqlitePortfolioStatsRepositoryImpl(db, objectMapper)
                val tradesRepo = SqliteTradeRepositoryImpl(db)

                val krakenService =
                    KrakenServiceImpl(
                        configService = mockConfigService,
                        objectMapper = objectMapper,
                        httpClient = httpClient,
                    )
                val portfolioAnalyzer =
                    PortfolioAnalyzerImpl(
                        krakenService = krakenService,
                        configService = mockConfigService,
                        portfolioStatsRepository = statsRepo,
                    )
                val tradeHistoryService =
                    TradeHistoryServiceImpl(
                        repository = tradesRepo,
                        portfolioStatsRepository = statsRepo,
                        krakenService = krakenService,
                        configService = mockConfigService,
                        objectMapper = objectMapper,
                        portfolioAnalyzer = portfolioAnalyzer,
                    )

                val orderExecutor =
                    OrderExecutorImpl(krakenService, tradeHistoryService)
                val portfolioManager =
                    PortfolioManagerImpl(
                        configService = mockConfigService,
                        tradeHistoryService = tradeHistoryService,
                        portfolioAnalyzer = portfolioAnalyzer,
                        orderExecutor = orderExecutor,
                    )

                portfolioManager.performRebalanceCycle()

                // 0.4 BTC @ $50,000 is $20,000 of a $50,000 portfolio (40%); restoring the 50/50 split
                // buys $5,000 of BTC, hence volume 0.1 on the XBTUSD pair alias.
                capturedOrderPayload.shouldNotBeNull()
                capturedOrderPayload.contains("pair=XBTUSD").shouldBeTrue()
                capturedOrderPayload.contains("type=buy").shouldBeTrue()
                capturedOrderPayload.contains("ordertype=market").shouldBeTrue()
                capturedOrderPayload.contains("volume=0.1").shouldBeTrue()
            }
        }
    }
}
